### PR TITLE
Fix: solve a empty order line with discount modification NaN error

### PR DESCRIPTION
**Pull request by jyling**
# Description

Please include a summary of the changes and the related issue.
This PR solves https://github.com/vendure-ecommerce/vendure/issues/3008, where there's an order level discount and we are updating the order to remove an order line by setting the quantity to 0. Currently, it throws a NaN error.

The solution is just to have a fallback of 0, when the `distribution[i]` returns undefined.

# Breaking changes

Does this PR include any breaking changes we should be aware of?
should not break existing functionality.

# Screenshots

### Note: I'm not sure why this test failed on my local machine, my changes shouldn't affect this test
image

# Checklist

📌 Always:
- [x] I have set a clear title
- [x] My PR is small and contains a single feature
- [x] I have [checked my own PR](## "Fix typo's and remove unused or commented out code")

👍 Most of the time:
- [x] I have added or updated test cases
- [x] I have updated the README if needed

### CODE REVIEW

**Comment by jyling** [comment]
I have read the CLA Document and I hereby sign the CLA

**Review by michaelbromley — CHANGES_REQUESTED** [review]
Thanks again - I reviewed this part of the code more closely and while you have indeed prevented the error from being thrown, I discovered another subtle bug that also needs to be addressed (my bug, not yours :D)

In the e2e test, if we add the following lines at the end, it will fail:

```ts
            // ensure correct adjustments applied
            // The first line should have a linePrice of 0 because it has zero quantity
            expect(modifyOrder.lines[0].linePriceWithTax).toBe(0);
            expect(modifyOrder.lines[0].proratedLinePriceWithTax).toBe(0);
            // The second line should have the proratedLinePriceWithTax discounted per the promotion
            expect(modifyOrder.lines[1].proratedLinePriceWithTax).toBe(
                modifyOrder.lines[1].discountedLinePriceWithTax / 2,
            );
```

The reason for this is subtle: on line 252 of `order-calculator.ts` we are filtering out zero-quantity order lines. This then means that the `weights` array (and the derived `distribution` array) will have fewer elements than the `order.lines` array.

So when on line 255 we loop over all the order lines, this is where we run into the error of trying to get an out-of-range index from the `distribution` array. Defaulting to `0` as in your fix will prevent the `NaN` error, but is still not correct, which is why the added e2e test assertions will fail.

The solution is to remove the `.filter()` call on line 252, to ensure that we do not end up with a smaller array and that `distribution.length === order.lines.length` at all times.

In summary, I recommend the following changes to make this fix correct:

1. Add the e2e assertions as above to your existing test
2. Remove the filter call on line 252
3. Change the map call to `.map(l => (l.quantity !== 0 ? l.proratedLinePriceWithTax : 0));`
4. At this point you can also remove the `?? 0` default since the `distribution` array will always have the correct length.

**Comment by jyling** [comment]
> In summary, I recommend the following changes to make this fix correct:

Thank you for the detailed explanation of the code that caused this bug, I will update the code and rerun the test.

**Comment by jyling** [comment]
I implemented the proposed fixes, which indeed solved the NaN issue, however i encountered another issue where the discounts is missing
image

image

I will look into this later, but just to update on this PR

**Comment by michaelbromley** [comment]
Tested locally and it looks good to me 👍
